### PR TITLE
Migrate DQMEDAnalyze to producers

### PR DESCRIPTION
**Pull request by dmitrijus**
Also fixes MEtoEDMConverter.

Supersedes https://github.com/cms-sw/cmssw/pull/22307

### CODE REVIEW

**Comment by davidlange6** [comment]
please test

**Comment by dmitrijus** [comment]
please test

File: DQMServices/Components/plugins/EDMtoMEConverter.h [hunk 1]
```diff
@@ -56,15 +56,16 @@ class EDMtoMEConverter : public edm::EDAnalyzer
   ~EDMtoMEConverter() override;
   void beginJob() override;
   void endJob() override;  
-  void analyze(const edm::Event&, const edm::EventSetup&) override;
   void beginRun(const edm::Run&, const edm::EventSetup&) override;
   void endRun(const edm::Run&, const edm::EventSetup&) override;
   void beginLuminosityBlock(const edm::LuminosityBlock&, const edm::EventSetup&) override;
   void endLuminosityBlock(const edm::LuminosityBlock&, const edm::EventSetup&) override;
-  void respondToOpenInputFile(const edm::FileBlock&) override;
+  void respondToOpenInputFile(const edm::FileBlock&) ;
```
Comment (fwyzard) [inline DQMServices/Components/plugins/EDMtoMEConverter.h]: if this is never called, you could simply drop it, as well as `iCountf`

File: DQMServices/Core/interface/DQMEDAnalyzer.h [hunk 2]
```diff
@@ -6,58 +6,39 @@
...
+                                                  edm::one::WatchRuns> 
 {
 public:
-  DQMEDAnalyzer() = default;
+  DQMEDAnalyzer();
   ~DQMEDAnalyzer() override = default;
   DQMEDAnalyzer(DQMEDAnalyzer const&) = delete;
   DQMEDAnalyzer(DQMEDAnalyzer &&) = delete;
 
-  void beginRun(edm::Run const& run, edm::EventSetup const& setup) final
```
Comment (fwyzard) [inline DQMServices/Core/interface/DQMEDAnalyzer.h]: same comment as in #22307: keep the trivial methods in the .h file

File: DQMServices/Components/plugins/EDMtoMEConverter.h (hunk 1 above)
Comment (dmitrijus) [reply DQMServices/Components/plugins/EDMtoMEConverter.h]: will be removed.

**Comment by dmitrijus** [comment]
@Dr15Jones I am using:

```
produces<DQMToken,edm::Transition::EndLuminosityBlock>();
produces<DQMToken,edm::Transition::EndRun>();
```

This gives me:

```
----- Begin Fatal Exception 23-Feb-2018 18:17:09 CET-----------------------
An exception of category 'LogicError' occurred while
   [0] Constructing the EventProcessor
   [1] Constructing module: class=TkAlCaRecoMonitor label='ALCARECOSiStripCalMinBiasAAGTrackerDQM'
Exception Message:
Duplicate Product The produced product Branch Type = LuminosityBlock
Process Name = RECO
ModuleLabel = ALCARECOSiStripCalMinBiasAAGTrackerDQM
Branch ID = 1450585091
Class Name = DQMToken
Friendly Class Name = DQMToken
Product Instance Name = 
 is registered more than once.
Please remove the redundant 'produces' call(s).
----- End Fatal Exception -------------------------------------------------
```

Should I go back to using two tokens: DQMToken and DQMRunToken? The code works fine with those.

If I give them labels, it works too:
```
produces<DQMToken,edm::Transition::EndLuminosityBlock>("endLumi");
produces<DQMToken,edm::Transition::EndRun>("endRun");
```

Can you confirm that consumesAll works fine with different labels?

**Comment by Dr15Jones** [comment]
@dmitrijus we think we know why you get the unexpected behavior. When telling an OutputModule what to `keep`, there is no way to specify if the item is in the Run or LuminosityBlock. So I believe the framework is requiring a unit branch name across Runs, LuminosityBlocks and Events. Not good, but that is what we have.

>Can you confirm that consumesAll works fine with different labels?

That is how the framework should work. I do not know if we've tested that explicitly.

**Comment by dmitrijus** [comment]
@Dr15Jones I have done the labels in the https://github.com/cms-sw/cmssw/pull/22319/commits/c27f399a3ed38527f897e585cade081bebd1b839

Should this PR should be ready to be reviewed and merged :)

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
The RelVal failures are do to a configuration error
```
----- Begin Fatal Exception 23-Feb-2018 21:05:57 CET-----------------------
An exception of category 'Configuration' occurred while
   [0] Constructing the EventProcessor
   [1] Validating configuration of module: class=OuterTrackerMonitorTTCluster label='OuterTrackerMonitorTTCluster'
Exception Message:
The base type in the python configuration is EDAnalyzer, but the base type
for the module's C++ class is EDProducer. Please fix the configuration.
It must use the same base type as the C++ class.
----- End Fatal Exception -------------------------------------------------
```

**Comment by dmitrijus** [comment]
please test

**Comment by dmitrijus** [comment]
Fixed the issue and rebased onto the latest IB.

**Comment by dmitrijus** [comment]
+1

File: DQMServices/Components/plugins/MEtoEDMConverter.cc [hunk 3]
```diff
@@ -92,274 +91,42 @@ void
 MEtoEDMConverter::beginJob()
 {
   // Determine if we are running multithreading asking to the DQMStore. Not to be moved in the ctor
+  DQMStore *dbe = edm::Service<DQMStore>().operator->();
   enableMultiThread_ = dbe->enableMultiThread_;
```
Comment (fwyzard) [inline DQMServices/Components/plugins/MEtoEDMConverter.cc]: isn't `enableMultiThread_` something you wanted to get rid of ?

File: DQMServices/Core/interface/DQMEDHarvester.h [hunk 4]
```diff
@@ -1,43 +1,45 @@
...
-//<<<<<< PUBLIC VARIABLES                                               >>>>>>
-//<<<<<< PUBLIC FUNCTIONS                                               >>>>>>
-//<<<<<< CLASS DECLARATIONS                                             >>>>>>
+#include "FWCore/Utilities/interface/EDPutToken.h"
+#include "DataFormats/Histograms/interface/DQMToken.h"
 
-class DQMEDHarvester
-: public edm::one::EDProducer<edm::one::WatchRuns,edm::one::WatchLuminosityBlocks,edm::one::SharedResources,
-edm::EndLuminosityBlockProducer>
+class DQMEDHarvester: public edm::one::EDProducer<edm::Accumulator,
```
Comment (fwyzard) [inline DQMServices/Core/interface/DQMEDHarvester.h]: why do you inherit from `edm::Accumulator`, if you do not read any event data ?

File: DQMServices/Core/src/DQMEDHarvester.cc [hunk 5]
```diff
@@ -8,24 +8,29 @@
...
   store->meBookerGetter([this, &iLumi, &iSetup](DQMStore::IBooker &b, DQMStore::IGetter &g){
-      this->dqmEndLuminosityBlock(b, g, iLumi, iSetup);
-    });
+    this->dqmEndLuminosityBlock(b, g, iLumi, iSetup);
+  });
 }
 
 void DQMEDHarvester::endLuminosityBlockProduce(edm::LuminosityBlock& iLumi, edm::EventSetup const&) {
-  iLumi.put(std::make_unique<DQMToken>());
+  iLumi.put(lumiToken_, std::make_unique<DQMToken>());
```
Comment (fwyzard) [inline DQMServices/Core/src/DQMEDHarvester.cc]: @Dr15Jones I'm not familiar with `EDPutToken`s... is it just an optimisation to use them, or do they enable any different behaviour ?

Comment (Dr15Jones) [reply DQMServices/Core/src/DQMEDHarvester.cc]: It is an optimization like `EDGetTokens`

File: DQMServices/Components/plugins/MEtoEDMConverter.cc (hunk 3 above)
Comment (dmitrijus) [reply DQMServices/Components/plugins/MEtoEDMConverter.cc]: It is. And we are getting rid of it. This PR is a step towards that - or rather to fix the https://github.com/cms-sw/cmssw/issues/22281 with least amount of changes.

File: DQMServices/Core/interface/DQMEDHarvester.h (hunk 4 above)
Comment (dmitrijus) [reply DQMServices/Core/interface/DQMEDHarvester.h]: The module has to implement either accumulate() or produce(), if none is implemented, it gives a compiler error.

Accumulate makes more sense.

File: DQMServices/Core/src/DQMStore.cc [hunk 6]
```diff
@@ -2070,6 +2025,43 @@ DQMStore::cloneLumiHistograms(uint32_t run, uint32_t lumi, uint32_t moduleId)
   }
 }
 
+/** Same as above, but for run histograms.
+ */
+
+void
+DQMStore::cloneRunHistograms(uint32_t run, uint32_t moduleId)
```
[the remark this answers is not in the document]
Comment (dmitrijus) [reply DQMServices/Core/src/DQMStore.cc]: The goal, again, was to restore the previous behaviour (pre-one module): all histograms are cloned.

Ideally, once we get rid of enableMultithread_ stuff, we can get rid away of this copying.

File: DQMServices/Core/src/DQMStore.cc [hunk 7]
```diff
@@ -395,6 +395,13 @@ DQMStore::DQMStore(const edm::ParameterSet &pset, edm::ActivityRegistry& ar)
         gc.luminosityBlockID().luminosityBlock(),
         mcc.moduleDescription()->id());
   });
```
[the remark this answers is not in the document]
Comment (dmitrijus) [reply DQMServices/Core/src/DQMStore.cc]: Good point and already changed.

**Comment by dmitrijus** [comment]
please test

File: DQMServices/Core/interface/DQMEDHarvester.h (hunk 4 above)
Comment (fwyzard) [reply DQMServices/Core/interface/DQMEDHarvester.h]: My understanding of what the framework will do is:
  - if standard `EDProducer` is not explicitly or implicitly scheduled, its `produce` method will not be called;
  - if an `EDProducer<Accumulator>` is not explicitly or implicitly scheduled, its `accumulate` method will be called for each event.

@Dr15Jones , sorry for bothering you, but did I understand correctly ?

Then, using a standard `EDProducer` for a harvesting module should be (slightly) more efficient.

File: DQMServices/Core/src/DQMStore.cc (hunk 6 above)
[the remark this answers is not in the document]
Comment (schneiml) [reply DQMServices/Core/src/DQMStore.cc]: #22307 now has re-written query logic, that should work without cloning.

However, I am not that sure if it does the right thing (by default MEs are per-run, but they can also be "global"), while maybe we want to have them global by default and only have per-run ones when needed. It does fix #22281 (even tough some of the changes of #22319 are missing over there), and also works in a number of other sequences, but might break things that I did not test.

**Comment by dmitrijus** [comment]
+1

File: DQMServices/Core/interface/DQMEDHarvester.h (hunk 4 above)
Comment (Dr15Jones) [reply DQMServices/Core/interface/DQMEDHarvester.h]: @fwyzard you are correct, the standard produce mechanism would be more efficient.

**Comment by Dr15Jones** [comment]
+1
removal of blank link in DataFormats/Histograms

**Comment by knoepfel** [comment]
@dmitrijus: congrats on much cleaner code--it is much clearer in terms of intent.  

I'm working on the concurrent booking of histograms in `DQMStore`, so I'd like to avoid collisions with you.  So far, so good, I think.  However, to support concurrent booking, the `std::set<MonitorElement>` data member will need to be replaced with something else that does not support the concept of `lower_bound`.  I'm not particularly concerned with that, but I'm giving you a heads-up.

**Comment by fwyzard** [comment]
Ha ha ha

**Comment by fwyzard** [comment]
(try not to break the DQM GUI, too)

**Comment by Dr15Jones** [comment]
@fwyzard see #18574

**Comment by fwyzard** [comment]
I would say that
  - we should not have two (groups of) people try to migrate the same code base at the same time
  - however the migration is done, it should be made sure that it does not break the use of the DQM code base outside of CMSSW

@dmitrijus and @schneiml should be working on a test to outline the set of necessary changes, possible steps, and required features.
I would suggest that @knoepfel discuss the possible changes with them beforehand ...

**Comment by schneiml** [comment]
@knoepfel yes, we should for sure have a talk at some point. We have a few plans on how to continue development -- and we can use any help we get.

**Comment by knoepfel** [comment]
@fwyzard: agreed--let's not collide/duplicate efforts.  

@schneiml, @dmitrijus: I can meet by Zoom, Skype, or whatever.  I can be available as early as 9 am (CST) tomorrow morning (2/28).  Otherwise, it will likely need to wait until next week.

**Comment by fabiocos** [comment]
+1

**Pull request merged**